File: library/src/main/java/com/lazygeniouz/house/ads/HouseAdsInterstitial.java
/*
 * Created by Darshan Pandya.
 * @itznotabug
 * Copyright (c) 2018.
 */

package com.lazygeniouz.house.ads;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.ActivityNotFoundException;
import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.os.Bundle;
import android.widget.ImageButton;
import android.widget.ImageView;

import com.lazygeniouz.house.ads.helper.JsonPullerTask;
import com.lazygeniouz.house.ads.listener.AdListener;
import com.lazygeniouz.house.ads.modal.InterstitialModal;
import com.squareup.picasso.Picasso;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;

import androidx.annotation.AnimRes;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

public class HouseAdsInterstitial {
    private final Context mContext;
    private static AdListener mAdListener;
    private String url;
    private int lastLoaded = 0;

    private static boolean isAdLoaded = false;
    private static Bitmap bitmap;
    private static String packageName;

    public HouseAdsInterstitial(Context context, String url) {
        this.mContext = context;
        this.url = url;
    }

    public void setAdListener(AdListener adListener) {
        mAdListener = adListener;
    }

    public void loadAd() {
        if (url.trim().isEmpty()) throw new IllegalArgumentException("Url is Blank!");
        else {
            new JsonPullerTask(url, result -> {
                if (!result.trim().isEmpty()) setUp(result);
                else {
                    if (mAdListener != null) mAdListener.onAdLoadFailed(new Exception("Null Response"));
                }
            }).execute();
        }
    }

    public boolean isAdLoaded() {
        return isAdLoaded;
    }

    @SuppressLint("CheckResult")
    private void setUp(String val) {
        ArrayList<InterstitialModal> modalArrayList = new ArrayList<>();
        String x = new String(new StringBuilder().append(val));

        try {
            JSONObject rootObject = new JSONObject(x);
            JSONArray array = rootObject.optJSONArray("apps");

            for (int object = 0; object < array.length(); object++) {
                JSONObject jsonObject = array.getJSONObject(object);

                if (jsonObject.optString("app_adType").equals("interstitial")) {
                    InterstitialModal interstitialModal = new InterstitialModal();
                    interstitialModal.setInterstitialImageUrl(jsonObject.optString("app_interstitial_url"));
                    interstitialModal.setPackageOrUrl(jsonObject.optString("app_uri"));
                    modalArrayList.add(interstitialModal);
                }
            }

        } catch (JSONException e) { e.printStackTrace(); }

        if (modalArrayList.size() > 0) {
            final InterstitialModal modal = modalArrayList.get(lastLoaded);
            if (lastLoaded == modalArrayList.size() - 1) lastLoaded = 0;
            else lastLoaded++;

            Picasso.get().load(modal.getInterstitialImageUrl()).into(new com.squareup.picasso.Target() {
                @Override
                public void onBitmapLoaded(Bitmap resource, Picasso.LoadedFrom from) {
                    bitmap = resource;
                    if (mAdListener != null) mAdListener.onAdLoaded();
                    isAdLoaded = true;
                }

                @Override
                public void onBitmapFailed(Exception e, Drawable errorDrawable) {
                    if (mAdListener != null) mAdListener.onAdLoadFailed(e);
                    isAdLoaded = false;
                }

                @Override
                public void onPrepareLoad(Drawable placeHolderDrawable) {

                }
            });
            packageName = modal.getPackageOrUrl();
        }
    }

    public void show() {
        mContext.startActivity(new Intent(mContext, InterstitialActivity.class));
        if (mContext instanceof AppCompatActivity) ((AppCompatActivity) mContext).overridePendingTransition(0, 0);
    }


    @SuppressWarnings("unused")
    public void show(@AnimRes int enterAnim, @AnimRes int exitAnim) {
        mContext.startActivity(new Intent(mContext, InterstitialActivity.class));
        if (mContext instanceof AppCompatActivity) ((AppCompatActivity) mContext).overridePendingTransition(enterAnim, exitAnim);
    }

    public static class InterstitialActivity extends Activity {

        @Override
        protected void onCreate(@Nullable Bundle savedInstanceState) {
            super.onCreate(savedInstanceState);
            if (mAdListener != null) mAdListener.onAdShown();

            setContentView(R.layout.house_ads_interstitial_layout);
            ImageView imageView = findViewById(R.id.image);
            ImageButton button = findViewById(R.id.button_close);

            imageView.setImageBitmap(bitmap);
            imageView.setOnClickListener(view -> {
                isAdLoaded = false;
                if (packageName.startsWith("http")) {
                    Intent val = new Intent(Intent.ACTION_VIEW, Uri.parse(packageName));
                    val.setPackage("com.android.chrome");
                    if (val.resolveActivity(getPackageManager()) != null) startActivity(val);
                    else startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse(packageName)));

                    if (mAdListener != null) mAdListener.onApplicationLeft();
                    finish();
                }
                else {
                    try {
                        startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse("market://details?id=" + packageName)));
                        if (mAdListener != null) mAdListener.onApplicationLeft();
                        finish();
                    } catch (ActivityNotFoundException e) {
                        startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse("http://play.google.com/store/apps/details?id=" + packageName)));
                        if (mAdListener != null) mAdListener.onApplicationLeft();
                        finish();
                    }
                }
            });
            button.setOnClickListener(view -> {
                finish();
                isAdLoaded = false;
                if (mAdListener != null) mAdListener.onAdClosed();
            });
        }

        @Override
        public void onBackPressed() {
            isAdLoaded = false;
            if (mAdListener != null) mAdListener.onAdClosed();
            finish();
        }
    }
}

File: library/src/main/java/com/lazygeniouz/house/ads/HouseAdsInterstitial.kt
/*
 * Created by Darshan Pandya. (@itznotabug)
 * Copyright (c) 2018-2020.
 */

@file:Suppress("unused")

package com.lazygeniouz.house.ads

import android.app.Activity
import android.content.ActivityNotFoundException
import android.content.Context
import android.content.Intent
import android.graphics.Bitmap
import android.graphics.Color
import android.graphics.PorterDuff
import android.graphics.drawable.BitmapDrawable
import android.graphics.drawable.GradientDrawable
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.ImageButton
import android.widget.ImageView
import android.widget.RelativeLayout
import androidx.annotation.AnimRes
import androidx.annotation.NonNull
import androidx.annotation.RawRes
import androidx.appcompat.app.AppCompatActivity
import androidx.cardview.widget.CardView
import coil.request.ErrorResult
import coil.request.SuccessResult
import com.lazygeniouz.house.ads.base.BaseAd
import com.lazygeniouz.house.ads.extension.*
import com.lazygeniouz.house.ads.helper.JsonHelper
import com.lazygeniouz.house.ads.listener.AdListener
import com.lazygeniouz.house.ads.modal.InterstitialModal
import org.json.JSONException
import org.json.JSONObject
import java.util.*

class HouseAdsInterstitial(context: Context, private val jsonUrl: String) : BaseAd(context) {

    private var isUsingRawRes = false
    private var jsonRawResponse = ""
    private var jsonLocalRawResponse = ""

    /**
     * Secondary constructor if you want to use a custom Json File from Raw folder
     */
    constructor(context: Context, @RawRes rawFile: Int) : this(context, "") {
        isUsingRawRes = true
        jsonLocalRawResponse = JsonHelper.getJsonFromRaw(context, rawFile)
    }

    /**
     * Set an [AdListener] to listen to Ad events
     *
     * Example: [AdListener.onAdLoaded], [AdListener.onAdFailedToLoad], etc
     */
    fun setAdListener(listener: AdListener): HouseAdsInterstitial {
        adListener = listener
        return this
    }

    /**
     * Set whether to use Palette API to color UI elements
     */
    fun usePalette(value: Boolean): HouseAdsInterstitial {
        usePalette = value
        return this
    }

    /**
     * Set whether to show / hide the system Navigation Bar
     * when the Interstitial Ad is displayed
     */
    fun hideNavigationBar(value: Boolean): HouseAdsInterstitial {
        hideNavigation = value
        return this
    }

    /**
     * Set the background & the icon color of the
     * Close Button **(X)** on the Interstitial Ad
     */
    fun setCloseButtonColor(
            backgroundColor: Int = Color.WHITE,
            iconTint: Int = Color.BLACK): HouseAdsInterstitial {
        closeButtonBackgroundColor = backgroundColor
        closeIconColor = iconTint
        return this
    }

    /**
     * Set the Close Button's **(X)** location
     *
     * Can either be [LOCATION.LOCATION_LEFT]
     *
     * or [LOCATION.LOCATION_RIGHT]
     */
    fun setCloseButtonLocation(location: LOCATION): HouseAdsInterstitial {
        closeButtonLocation = location
        return this
    }

    /**
     * Google AdMob has recently applied this
     * behavior to their Interstitial Ads.
     *
     * If **true**, Interstitial Ad can be closed by a BackPress.
     * Else the user will have to click on the Close Button **(X)**
     */
    fun exitOnBackPress(exit: Boolean): HouseAdsInterstitial {
        exitOnBackPress = exit
        return this
    }

    /**
     * Alright, lets load the Ads
     */
    fun loadAds() {
        isAdLoaded = false
        if (!isUsingRawRes) {
            require(jsonUrl.trim().isNotEmpty()) { context.getString(R.string.error_url_blank) }
            if (jsonRawResponse.isEmpty()) {
                launch {
                    val result = JsonHelper.getJsonObject(jsonUrl)
                    if (result.trim().isNotEmpty()) {
                        jsonRawResponse = result
                        configureAds(result)
                    } else adListener?.onAdFailedToLoad(Exception(context.getString(R.string.error_null_response)))
                }
            } else configureAds(jsonRawResponse)
        } else configureAds(jsonLocalRawResponse)
    }

    /**
     * Show the loaded Ad
     */
    fun show() {
        context.startActivity(Intent(context, InterstitialActivity::class.java))
        if (context is AppCompatActivity) context.overridePendingTransition(0, 0)
    }

    /**
     * Show the loaded Ad with activity transitions
     */
    fun show(@NonNull @AnimRes enterAnim: Int, @NonNull @AnimRes exitAnim: Int) {
        context.startActivity(Intent(context, InterstitialActivity::class.java))
        if (context is Activity) context.overridePendingTransition(enterAnim, exitAnim)
        else Log.d(TAG, "show(enterAnim, exitAnim) cannot be used because the Context is not an instance of Activity")
    }

    /**
     * Check if the Interstitial Ad is loaded
     */
    fun isAdLoaded(): Boolean {
        return isAdLoaded
    }

    private fun configureAds(jsonResponse: String) {
        val modalArrayList = ArrayList<InterstitialModal>()

        try {
            val rootObject = JSONObject(jsonResponse)
            val array = rootObject.optJSONArray("apps")!!
            for (childObject in 0 until array.length()) {
                val jsonObject = array.getJSONObject(childObject)
                if (jsonObject.optString("app_adType") == "interstitial")
                    modalArrayList.add(getInterstitialModal(jsonObject))
            }
        } catch (jsonException: JSONException) {
            jsonException.printStackTrace()
        }

        val modal = modalArrayList[lastLoaded]
        if (modalArrayList.size > 0) {
            if (lastLoaded == modalArrayList.size - 1) lastLoaded = 0
            else lastLoaded++

            if (!isUsingRawRes) {
                require(!(modal.interstitialImageUrl.isEmpty() || !modal.interstitialImageUrl.hasHttpSign))
                { context.getString(R.string.error_interstitial_image_url_null) }
            } else {
                when {
                    modal.interstitialImageUrl.trim().hasHttpSign -> Log.d(TAG, "ImageUrl starts with `http://`")
                    modal.interstitialImageUrl.trim().hasDrawableSign -> Log.d(TAG, "ImageUrl is a local drawable")
                    else -> throw IllegalArgumentException(context.getString(R.string.error_raw_resource_interstitial_image_null))
                }
            }

            val imageUrlToLoad: String = if (modal.interstitialImageUrl.hasDrawableSign) {
                context.getDrawableUriAsString(modal.interstitialImageUrl)
            } else modal.interstitialImageUrl

            launch {
                when (val result = getImageFromNetwork(imageUrlToLoad)) {
                    is SuccessResult -> {
                        bitmap = (result.drawable as BitmapDrawable).bitmap
                        if (usePalette && bitmap != null) paletteColor = bitmap!!.getDominantColorForInterstitial()
                        adListener?.onAdLoaded()
                        isAdLoaded = true
                    }
                    is ErrorResult -> {
                        adListener?.onAdFailedToLoad(Exception(result.throwable))
                        isAdLoaded = false
                    }
                }
                packageName = modal.packageOrUrl
            }
        }
    }

    class InterstitialActivity : Activity() {
        override fun onCreate(savedInstanceState: Bundle?) {
            fullscreen()
            super.onCreate(savedInstanceState)

            adListener?.onAdShown()
            setContentView(R.layout.house_ads_interstitial_layout)
            val parent = findViewById<RelativeLayout>(R.id.houseAds_interstitial_parent)
            val imageView = findViewById<ImageView>(R.id.houseAds_interstitial_image)
            val cardView = findViewById<CardView>(R.id.houseAds_interstitial_button_close_card)
            val button = findViewById<ImageButton>(R.id.houseAds_interstitial_button_close)

            parent.setBackgroundColor(paletteColor)
            cardView.apply {
                if (closeButtonLocation == LOCATION.LOCATION_RIGHT)
                    layoutParams = (layoutParams as RelativeLayout.LayoutParams)
                            .apply {
                                addRule(RelativeLayout.ALIGN_PARENT_END)
                            }
            }

            imageView.apply {
                setImageBitmap(bitmap)
                setOnClickListener {
                    isAdLoaded = false
                    if (packageName!!.hasHttpSign) {
                        val packageIntent = Intent(Intent.ACTION_VIEW, Uri.parse(packageName))
                        packageIntent.setPackage("com.android.chrome")
                        if (packageIntent.resolveActivity(packageManager) != null)
                            startActivity(packageIntent)
                        else startActivity(Intent(Intent.ACTION_VIEW, Uri.parse(packageName)))

                        adListener?.onApplicationLeft()
                        finish()
                    } else {
                        try {
                            startActivity(Intent(Intent.ACTION_VIEW, Uri.parse("market://details?id=" + packageName!!)))
                            adListener?.onApplicationLeft()
                            finish()
                        } catch (e: ActivityNotFoundException) {
                            startActivity(Intent(Intent.ACTION_VIEW, Uri.parse("http://play.google.com/store/apps/details?id=" + packageName!!)))
                            adListener?.onApplicationLeft()
                            finish()
                        }
                    }
                }
            }

            button.apply {
                background = GradientDrawable().apply {
                    shape = GradientDrawable.OVAL
                    cornerRadii = floatArrayOf(0f, 0f, 0f, 0f, 0f, 0f, 0f, 0f)
                    setColor(closeButtonBackgroundColor)
                }
                setColorFilter(closeIconColor, PorterDuff.Mode.SRC_ATOP)
                setOnClickListener {
                    finish()
                    isAdLoaded = false
                    adListener?.onAdClosed()
                }
            }
        }

        override fun onBackPressed() {
            if (exitOnBackPress) {
                isAdLoaded = false
                adListener?.onAdClosed()
                finish()
            }
        }

        private fun fullscreen() {
            if (Build.VERSION.SDK_INT >= 19) {
                val decorView: View = window.decorView
                val uiOptions: Int = if (hideNavigation) View.SYSTEM_UI_FLAG_HIDE_NAVIGATION or View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY
                else View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY
                decorView.systemUiVisibility = uiOptions
            }
        }
    }

    // to be used inside the nested activity
    private companion object {
        private var lastLoaded = 0
        private var adListener: AdListener? = null
        private var isAdLoaded = false
        private var bitmap: Bitmap? = null
        private var exitOnBackPress: Boolean = false
        private var paletteColor: Int = Color.BLACK
        private var closeButtonBackgroundColor = Color.WHITE
        private var closeButtonLocation = LOCATION.LOCATION_LEFT
        private var closeIconColor = Color.BLACK
        private var usePalette: Boolean = false
        private var hideNavigation: Boolean = true
        private var packageName: String? = null
        private val TAG = HouseAdsInterstitial::class.java.simpleName
    }

    enum class LOCATION {
        LOCATION_LEFT,
        LOCATION_RIGHT
    }
}
